src/security/DigestAlgorithm.h

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright (C) 1996-2018 The Squid Software Foundation and contributors
+ *
+ * Squid software is distributed under GPLv2+ license and includes
+ * contributions from numerous individuals and organizations.
+ * Please see the COPYING and CONTRIBUTORS files for details.
+ */
+
+#ifndef SQUID__SRC_SECURITY_DIGESTALGORITHM_H
+#define SQUID__SRC_SECURITY_DIGESTALGORITHM_H
+
+namespace Security
+{
+
+#if USE_OPENSSL
+typedef EVP_MD const * DigestAlgorithm;
+#elif USE_GNUTLS
+typedef gnutls_digest_algorithm_t DigestAlgorithm;
+#else
+typedef void * DigestAlgorithm;
```

Comment:
IIRC, we found a better way to represent pointers that should never be dereferenced but may need default initialization values to prevent compiler warnings. Please use that trick here if it is applicable.

Reply:
are you referring to "class Foo {}" definition?

Reply:
AFAICT, the current always-nil should-not-be-dereferenced pointer definition is `std::shared_ptr<void>` (used by CertPointer and PrivateKeyPointer).

Introducing a dedicated class for the same purpose would be better than using one of the standard smart pointers if (and only if) std::shared_ptr, unique_ptr, and all other standard smart pointers come with some significant performance overhead or other problems. I am not sure they do, but I have not researched it (or forgot the results of that research).

Note that an empty class may not work well: Checking the pointer for being nil is, arguably, a valid operation for all these `#else` pointers. A smart void pointer supports that check but an empty class does not (which can be fixed by adding a method or two). Similar logic applies to printing these pointers to std::ostream for debugging.

Reply:
and reverted due to followup request (other DigestAlgorithm types are raw-pointer so making this a smart pointer is wrong).